File: frontend/src/lib/tauri/yearly-bingo.ts
// Yearly Bingo — yearly_bingo.rs. Only the board crosses IPC (its size and
// one snapshot of the picked work per cell); results are computed from the
// cached library in lib/bingo/bingo-result.ts.
import { invoke, tauriCmd } from './bridge';

export interface BingoItem {
  external_id: string;
  title: string;
  cover_url: string | null;
  media_type: string;
}

/** null = empty cell. */
export type BingoCell = BingoItem | null;

export interface YearlyBingoData {
  year: number;
  /** Number of cells the owner picked, 1–49 (boards saved before sizes existed read as 16). */
  size: number;
  /** Exactly `size` cells, in grid order. */
  items: BingoCell[];
  /** Unix seconds. */
  created_at: number;
  updated_at: number;
}

/** `year`'s board, or null when none was saved. Rejects on an IPC/DB error
 *  so the UI never offers to overwrite a board it could not read. Outside
 *  Tauri: null. */
export function getYearlyBingo(year: number): Promise<YearlyBingoData | null> {
  return tauriCmd<YearlyBingoData | null>('get_yearly_bingo', null, { year });
}

/** Replaces `year`'s board (all cells empty deletes it). `items` must hold
 *  exactly `size` (1–49) cells, else E_BINGO_INVALID; Rust rejects it with
 *  E_BINGO_LOCKED outside the Dec 20 – Jan 10 edit window. */
export function setYearlyBingo(year: number, size: number, items: BingoCell[]): Promise<void> {
  return invoke<void>('set_yearly_bingo', { year, size, items });
}

File: frontend/src-tauri/src/migrations/yearly_bingo.rs
//! Yearly Bingo replaces the "Challenge {year}" tab (migration 79's goals +
//! resolutions, and the social_* mirrors migration 80 added for visited
//! profiles) — those tables are dropped with their rows.
//! - yearly_bingo: one board per year (yearly_bingo.rs). `items` is a JSON
//!   array of 1–49 cells (its length is the board's size, 16 by default),
//!   in grid order, each null (empty) or a snapshot
//!   {external_id, title, cover_url, media_type} taken when the work was
//!   picked. Not a foreign key: a work needn't be in the library (or even
//!   cached in media_catalog). The edit lock (Jan 10 of `year`, local date)
//!   is computed, not stored. created_at / updated_at are unix seconds.
use rusqlite::{Result as SqlResult, Transaction};

pub(super) fn migrate(tx: &Transaction) -> SqlResult<()> {
    tx.execute_batch(
        "DROP TABLE IF EXISTS yearly_goals;
        DROP TABLE IF EXISTS yearly_resolutions;
        DROP TABLE IF EXISTS yearly_resolution_settings;
        DROP TABLE IF EXISTS social_yearly_goals;
        DROP TABLE IF EXISTS social_yearly_resolutions;
        DROP TABLE IF EXISTS social_yearly_resolution_settings;
        CREATE TABLE IF NOT EXISTS yearly_bingo (
            year       INTEGER PRIMARY KEY,
            items      TEXT    NOT NULL,
            created_at INTEGER NOT NULL,
            updated_at INTEGER NOT NULL
        );",
    )
}

#[cfg(test)]
mod tests {
    use rusqlite::Connection;

    #[test]
    fn drops_the_challenge_tables_and_creates_the_board_table() {
        let mut conn = Connection::open_in_memory().unwrap();
        conn.execute_batch(
            "CREATE TABLE yearly_goals (year INTEGER, media_type TEXT, target INTEGER);
            INSERT INTO yearly_goals VALUES (2026, 'book', 12);
            CREATE TABLE social_yearly_resolutions (social_user_id TEXT, year INTEGER, external_id TEXT);",
        )
        .unwrap();
        let tx = conn.transaction().unwrap();
        super::migrate(&tx).unwrap();
        tx.commit().unwrap();
        let tables: Vec<String> = conn
            .prepare("SELECT name FROM sqlite_master WHERE type = 'table' ORDER BY name")
            .unwrap()
            .query_map([], |r| r.get(0))
            .unwrap()
            .collect::<Result<_, _>>()
            .unwrap();
        assert_eq!(tables, vec!["yearly_bingo".to_string()]);
    }
}
